oprf/group/suite.go

```diff
@@ -0,0 +1,251 @@
+package group
+
+import (
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/binary"
+	"errors"
+	"hash"
+	"io"
+	"math/big"
+
+	"github.com/armfazh/h2c-go-ref"
+)
+
+// Ciphersuite corresponds to the OPRF ciphersuite that is chosen. The
+// Ciphersuite object determines the prime-order group (pog) that is used for
+// performing the OPRF operations, along with the different hash function
+// definitions.
+// Should be created using NewSuite, using an string.
+type Ciphersuite struct {
+	// name of the ciphersuite.
+	name string
+
+	// dst is a tag to be used for hashing to curve.
+	dst []byte
+
+	// hash defines the hash function to be used.
+	hash hash.Hash
+
+	curve elliptic.Curve
+}
+
+// Suite is an interface that defines operations within additive
+// groups of prime order. This is the setting in which the OPRF operations
+// take place.
+type Suite interface {
+	// Returns the identifying name of the group
+	Name() string
+
+	// Returns the canonical (fixed) generator for defined group
+	Generator() *Element
+
+	// Returns the order of the canonical generator in the group.
+	Order() *Scalar
+
+	// Performs a transformation to encode bytes as a Element object in the
+	// group.
+	HashToGroup([]byte) (*Element, error)
+
+	// Performs a transformation to encode bytes as a Scalar object in the
+	// appropriate group.
+	HashToScalar([]byte) (Scalar, error)
+
+	// Samples a random scalar value from the field of scalars defined by the
+	// group order.
+	RandomScalar() *Scalar
+
+	// Performs a scalar multiplication of the Generator with some scalar
+	// input
+	ScalarMultBase(*Scalar) *Element
+
+	// Returns the ByteLength of objects associated with the Ciphersuite
```

Comment:
What is an "object" here? An element? Scalar? Something else?

Reply:
Clarified now.